aliyun-ecs/src/test/java/org/jclouds/aliyun/ecs/compute/functions/InstanceToNodeMetadataTest.java

```diff
@@ -0,0 +1,214 @@
+              .expiredTime(new SimpleDateFormatDateService().iso8601DateParse("2014-03-22T09:16:45.784120972Z"))
+              .innerIpAddress(ImmutableMap.<String, List<String>>of("IpAddress", ImmutableList.of("192.168.0.1")))
+              .publicIpAddress(ImmutableMap.<String, List<String>>of("IpAddress", ImmutableList.of("47.254.152.220")))
+              .securityGroupIds(ImmutableMap.<String, List<String>>of())
+              .networkInterfaces(ImmutableMap.<String, List<NetworkInterface>>of())
+              .operationLocks(ImmutableMap.<String, List<String>>of())
+              .tags(ImmutableMap.<String, List<Tag>>of())
+              .build();
+
+      Supplier<Map<String, ? extends Image>> images = new Supplier<Map<String, ? extends Image>>() {
+         @Override
+         public Map<String, ? extends Image> get() {
+            return ImmutableMap.of("centos_6_09_64_20G_alibase_20180326.vhd", image);
+         }
+      };
+
+      Supplier<Map<String, ? extends Hardware>> hardwares = new Supplier<Map<String, ? extends Hardware>>() {
+         @Override
+         public Map<String, ? extends Hardware> get() {
+            return ImmutableMap.of("ecs.t1.xsmall", hardware);
+         }
+      };
+
+      instanceToNodeMetadata = new InstanceToNodeMetadata(images, hardwares, locations,
+              instanceStatusToStatus, conventionFactory);
+   }
+
+   @Test
+   public void testApply() {
+      expect(nodeNamingConvention.extractGroup(serverName)).andReturn("[" + serverName + "]").anyTimes();
+      expect(instanceStatusToStatus.apply(anyObject(Instance.Status.class))).andReturn(NodeMetadata.Status.RUNNING);
+
+      EasyMock.replay(nodeNamingConvention, instanceStatusToStatus);
+
+      assertNodeMetadata(instanceToNodeMetadata.apply(instance), os,
+            instance.imageId(), NodeMetadata.Status.RUNNING,
+            instance.innerIpAddress().entrySet().iterator().next().getValue(),
+            instance.publicIpAddress().entrySet().iterator().next().getValue());
+   }
+
+   private void assertNodeMetadata(NodeMetadata result, org.jclouds.compute.domain.OperatingSystem os, String imageId,
+                                   NodeMetadata.Status status, List<String> privateIpAddresses, List<String> publicIpAddresses) {
+      assertNotNull(result);
+      assertEquals(result.getProviderId(), instance.id());
+      assertEquals(result.getName(), instance.name());
+      assertEquals(result.getHostname(), instance.hostname());
+      assertEquals(result.getGroup(), "[" + serverName + "]");
+      assertEquals(result.getHardware(), hardware);
+      assertEquals(result.getOperatingSystem(), os);
+      assertEquals(result.getLocation(), location);
+      assertEquals(result.getImageId(), RegionAndId.slashEncodeRegionAndId(regionId, imageId));
+      assertEquals(result.getStatus(), status);
+      assertEquals(result.getPrivateAddresses(), privateIpAddresses);
+      assertEquals(result.getPublicAddresses(), publicIpAddresses);
```

Comment:
There is custom logic to extract the tags. Make sure you test tags too.
Also add a couple tests to verify what happens when the image and hardware are not found.